Suboptimal debug codegen for integer division with a constant rhs

Codegen for division in general needs to generate code to panic if the division will panic (e.g. divide by zero or integer overflow). In the case of constants, however, it looks like rustc *almost* optimizes debug codegen but not quite. For example this code:

```rust
#[no_mangle]
pub extern fn foo(a: i32) -> i32 {
     a / 256
}
```

[generates this IR](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=754a67acfbac34f972423f4b28fa4da2) in debug mode:

```llvm
define i32 @foo(i32 %a) unnamed_addr #0 !dbg !6 {
start:
  %a.dbg.spill = alloca i32, align 4
  store i32 %a, i32* %a.dbg.spill, align 4
  call void @llvm.dbg.declare(metadata i32* %a.dbg.spill, metadata !12, metadata !DIExpression()), !dbg !13
  %_4 = icmp eq i32 %a, -2147483648, !dbg !14
  %_5 = and i1 false, %_4, !dbg !14
  %0 = call i1 @llvm.expect.i1(i1 %_5, i1 false), !dbg !14
  br i1 %0, label %panic, label %bb1, !dbg !14

bb1:                                              ; preds = %start
  %1 = sdiv i32 %a, 256, !dbg !14
  ret i32 %1, !dbg !15

panic:                                            ; preds = %start
; call core::panicking::panic
  call void @_ZN4core9panicking5panic17h8a9eda1e10298363E([0 x i8]* noalias nonnull readonly align 1 bitcast ([31 x i8]* @str.0 to [0 x i8]*), i64 31, %"core::panic::Location"* noalias readonly align 8 dereferenceable(24) bitcast (<{ i8*, [16 x i8] }>* @anon.670bbfb7b436af44f45c978390c94a0e.1 to %"core::panic::Location"*)), !dbg !14
  unreachable, !dbg !14
}
```

It looks like rustc is "smart enough" to generate `%_5 = and i1 false, %_4, !dbg !14` to realize the rhs is not negative one and it's not zero, but it's not quite smart enough to know that `false && x` is always `false`, so it still generates a branch to panic.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"saethlin"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_mir_transform/src/const_prop.rs
+++ b/compiler/rustc_mir_transform/src/const_prop.rs
@@ -655,11 +655,7 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             return None;
         }
 
-        if self.tcx.sess.mir_opt_level() >= 4 {
-            self.eval_rvalue_with_identities(rvalue, place)
-        } else {
-            self.use_ecx(|this| this.ecx.eval_rvalue_into_place(rvalue, place))
-        }
+        self.eval_rvalue_with_identities(rvalue, place)
     }
 
     // Attempt to use algebraic identities to eliminate constant expressions

--- a/src/test/mir-opt/div_overflow.const_dividend.PreCodegen.after.mir
+++ b/src/test/mir-opt/div_overflow.const_dividend.PreCodegen.after.mir
@@ -1,0 +1,17 @@
+// MIR for `const_dividend` after PreCodegen
+
+fn const_dividend(_1: i32) -> i32 {
+    debug a => _1;                       // in scope 0 at $DIR/div_overflow.rs:+0:23: +0:24
+    let mut _0: i32;                     // return place in scope 0 at $DIR/div_overflow.rs:+0:34: +0:37
+    let mut _2: bool;                    // in scope 0 at $DIR/div_overflow.rs:+1:5: +1:12
+
+    bb0: {
+        _2 = Eq(_1, const 0_i32);        // scope 0 at $DIR/div_overflow.rs:+1:5: +1:12
+        assert(!move _2, "attempt to divide `{}` by zero", const 256_i32) -> bb1; // scope 0 at $DIR/div_overflow.rs:+1:5: +1:12
+    }
+
+    bb1: {
+        _0 = Div(const 256_i32, move _1); // scope 0 at $DIR/div_overflow.rs:+1:5: +1:12
+        return;                          // scope 0 at $DIR/div_overflow.rs:+2:2: +2:2
+    }
+}

--- a/src/test/mir-opt/div_overflow.const_divisor.PreCodegen.after.mir
+++ b/src/test/mir-opt/div_overflow.const_divisor.PreCodegen.after.mir
@@ -1,0 +1,11 @@
+// MIR for `const_divisor` after PreCodegen
+
+fn const_divisor(_1: i32) -> i32 {
+    debug a => _1;                       // in scope 0 at $DIR/div_overflow.rs:+0:22: +0:23
+    let mut _0: i32;                     // return place in scope 0 at $DIR/div_overflow.rs:+0:33: +0:36
+
+    bb0: {
+        _0 = Div(move _1, const 256_i32); // scope 0 at $DIR/div_overflow.rs:+1:5: +1:12
+        return;                          // scope 0 at $DIR/div_overflow.rs:+2:2: +2:2
+    }
+}

--- a/src/test/mir-opt/div_overflow.rs
+++ b/src/test/mir-opt/div_overflow.rs
@@ -1,0 +1,18 @@
+// compile-flags: -Copt-level=0 -Coverflow-checks=yes
+
+// Tests that division with a const does not emit a panicking branch for overflow
+
+// EMIT_MIR div_overflow.const_divisor.PreCodegen.after.mir
+pub fn const_divisor(a: i32) -> i32 {
+    a / 256
+}
+
+// EMIT_MIR div_overflow.const_dividend.PreCodegen.after.mir
+pub fn const_dividend(a: i32) -> i32 {
+    256 / a
+}
+
+fn main() {
+    const_divisor(123);
+    const_dividend(123);
+}

